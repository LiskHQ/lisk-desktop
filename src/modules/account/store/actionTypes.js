const actionTypes = {
  setCurrentAccount: 'SET_CURRENT_ACCOUNT',
  updateCurrentAccount: 'UPDATE_CURRENT_ACCOUNT',
  addAccount: 'ADD_ACCOUNT',
  updateAccount: 'UPDATE_ACCOUNT',
  deleteAccount: 'DELETE_ACCOUNT',
};

export default actionTypes;
