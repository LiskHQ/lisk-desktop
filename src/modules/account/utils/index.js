export * from './encryptAccount';
export * from './decryptAccount';
