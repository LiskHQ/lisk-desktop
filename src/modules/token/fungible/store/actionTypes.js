const actionTypes = {
  tokenReset: 'TOKEN_RESET',
  tokenUpdated: 'TOKEN_UPDATED',
  tokenRetrieved: 'TOKEN_RETRIEVED',
};

export default actionTypes;
