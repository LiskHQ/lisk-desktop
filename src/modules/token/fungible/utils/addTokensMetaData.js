export const addTokensMetaData =
  ({ createMetaConfig, client }) =>
  async (tokens) => {
    try {
      const tokenIDs = tokens?.map(({ tokenID }) => tokenID).join(',');
      if (!tokenIDs) {
        return tokens;
      }
      const metaConfig = createMetaConfig({
        params: {
          tokenID: tokenIDs,
          limit: tokens.length,
          chainID: null, // Omit chainID while fetching all token metadata for a given network
        },
      });
      const metaRes = await client.call(metaConfig);
      return tokens.map((token) => {
        const selectedTokenMeta = metaRes?.data?.find(
          (tokenMeta) => tokenMeta.tokenID === token.tokenID
        );
        return { ...(selectedTokenMeta ?? {}), ...token };
      });
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log('meta tokens Error:', e);
      return tokens;
    }
  };
