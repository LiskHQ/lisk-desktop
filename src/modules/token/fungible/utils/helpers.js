import BigNumber from 'bignumber.js';
import liskLogo from '../../../../../setup/react/assets/images/LISK.png';

export const getLogo = ({ logo }) => logo?.svg || logo?.png || liskLogo;

BigNumber.config({ ERRORS: false });
const BASE = 10;

const defaultTokenMeta = {
  displayDenom: 'lsk',
  denomUnits: [
    {
      denom: 'beddows',
      decimals: 0,
      aliases: ['Beddows'],
    },
    {
      denom: 'lsk',
      decimals: 8,
      aliases: ['Lisk'],
    },
  ],
};

export const getTokenDecimals = (token) => {
  const { decimals } = token.denomUnits?.find?.(({ denom }) => denom === token.displayDenom) || {};

  return decimals;
};

/**
 * Converts a given token amount to its token symbol denom
 */
export const convertFromBaseDenom = (amount, token = defaultTokenMeta) => {
  const decimals = getTokenDecimals(token);

  if (!decimals) return '0';

  return new BigNumber(amount || 0).dividedBy(new BigNumber(BASE).pow(decimals)).toFixed();
};

/**
 * Converts from denom denoted by the token symbol to a smaller unit
 */
export const convertToBaseDenom = (amount, token = defaultTokenMeta) => {
  // eslint-disable-next-line no-restricted-globals
  if (isNaN(amount)) return '0';

  const decimals = getTokenDecimals(token);

  if (!decimals) return '0';

  return new BigNumber(amount || 0).multipliedBy(new BigNumber(BASE).pow(decimals)).toFixed();
};
