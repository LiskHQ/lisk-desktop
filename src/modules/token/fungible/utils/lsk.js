import BigNumber from 'bignumber.js';
import numeral from 'numeral';
import 'numeral/locales';

BigNumber.config({ ERRORS: false });

/**
 * Convert Beddow to LSK
 *
 * @param {Strong|Number} value - Value in Beddow
 * @returns {BigNumber} Value converted to LSK
 */
export const fromRawLsk = (value) =>
  new BigNumber(value || 0).dividedBy(new BigNumber(10).pow(8)).toFixed();

/**
 * Converts a given token amount to its token symbol denom
 *
 * @param {BigNumber|number} amount - Amount value to be converted
 * @param {BigNumber|number} token - Token value merged with its equivalent metadata
 * @returns {BigNumber} Amount value converted to the token symbol's denom
 */
export const changeDemnom = (amount, token = {}) => {
  const { decimals = 8 } =
    token.denomUnits?.find?.(({ denom }) => denom === token.symbol.toLowerCase()) || {};

  return new BigNumber(amount || 0).dividedBy(new BigNumber(10).pow(decimals)).toFixed();
};

/**
 * Convert LSK to Beddow
 *
 * @param {Strong|Number} value - Value in LSK
 * @returns {BigNumber} Value converted to Beddow
 */
export const toRawLsk = (value) => {
  const amount = numeral(value).value();
  return new BigNumber(amount * new BigNumber(10).pow(8)).decimalPlaces(0).toNumber();
};

/**
 * After a new block is created and broadcasted
 * it takes a few ms for Lisk Service
 * to update transactions index, so we need to wait
 * before retrieving the the transaction by blockId
 *
 * @returns {Promise} resolves with True after 100ms
 */
/* istanbul ignore next */
export const delay = (ms = 1500) =>
  new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, ms);
  });
