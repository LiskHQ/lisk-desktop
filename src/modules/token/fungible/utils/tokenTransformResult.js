import { clientMetaData } from 'src/utils/api/client';

export const tokenTransformResult = ({createMetaConfig}) => async (res) => {
  const tokenIDs = res?.data?.map(({ tokenID }) => tokenID).join(',');
  if (!tokenIDs) {
    return res;
  }
  const metaConfig = createMetaConfig({ params: { tokenID: tokenIDs, limit: 100 } });
  const metaRes = await clientMetaData.call(metaConfig);
  return {
    ...res,
    data: res?.data?.map((token) => {
      const selectedTokenMeta = metaRes?.data?.find(
        (tokenMeta) => tokenMeta.tokenID === token.tokenID
      ) ?? {};
      return { ...selectedTokenMeta, ...token};
    }),
  };
};
