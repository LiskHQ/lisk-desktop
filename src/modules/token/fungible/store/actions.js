// @todo: this should be re-instated when the issue with lisk-client is fixed
/* istanbul ignore file */
import { to } from 'await-to-js';
import { selectActiveTokenAccount } from 'src/redux/selectors';
import actionTypes from '@transaction/store/actionTypes';
import { createGenericTx } from '@transaction/api/index';

/**
 * Calls transactionAPI.create for create the tx object that will broadcast
 * @param {Object} data
 * @param {String} data.recipientAddress
 * @param {Number} data.amount - In raw format (beddows)
 * @param {Number} data.fee - In raw format, used for updating the TX List.
 * @param {Number} data.reference - Data field for LSK transactions
 */

export const tokensTransferred = (
  transactionObject,
  privateKey,
) => async (dispatch, getState) => {
  const state = getState();
  const wallet = selectActiveTokenAccount(state);

  const [error, tx] = await to(
    createGenericTx({
      transactionObject,
      wallet,
      schema: state.network.networks.LSK.moduleCommandSchemas[transactionObject.moduleCommand],
      chainID: state.network.networks.LSK.chainID,
      privateKey,
    }),
  );

  if (error) {
    dispatch({
      type: actionTypes.transactionSignError,
      data: error,
    });
  } else {
    dispatch({
      type: actionTypes.transactionCreatedSuccess,
      data: tx,
    });
  }
};
