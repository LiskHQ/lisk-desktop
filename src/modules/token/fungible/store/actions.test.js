import { cryptography } from '@liskhq/lisk-client';
import loginTypes from '@auth/const/loginTypes';
import * as hwManager from '@transaction/utils/hwManager';
import { getState } from '@tests/fixtures/transactions';
import actionTypes from '@transaction/store/actionTypes';
import wallets from '@tests/constants/wallets';
import { getAddressFromBase32Address } from '@wallet/utils/account';
import { convertStringToBinary } from '@transaction/utils/transaction';
import { tokensTransferred } from './actions';

jest.mock('@transaction/utils/hwManager');

describe('actions: transactions', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  const dispatch = jest.fn();

  describe('tokensTransferred', () => {
    const state = getState();

    const activeAccount = {
      ...state.wallet.info.LSK,
      hwInfo: {
        deviceModel: 'Ledger Nano S',
      },
      passphrase: state.wallet.passphrase,
    };
    const getStateWithHW = () => ({
      ...state,
      wallet: {
        info: {
          LSK: activeAccount,
        },
        hwInfo: {
          deviceModel: 'Ledger Nano S',
        },
        passphrase: state.wallet.passphrase,
      },
    });

    it('should dispatch tokensTransferredSuccess action if there are no errors', async () => {
      // Arrange
      const data = {
        fee: 141000,
        moduleCommandID: '2:0',
        sender: {
          publicKey: wallets.genesis.summary.publicKey,
        },
        nonce: '2',
        params: {
          recipient: { address: wallets.genesis.summary.address },
          amount: 112300000,
          data: 'test',
        },
      };
      const tx = {
        fee: BigInt(141000),
        moduleID: cryptography.utils.intToBuffer(2, 4),
        commandID: cryptography.utils.intToBuffer(0, 4),
        senderPublicKey: convertStringToBinary(wallets.genesis.summary.publicKey),
        nonce: BigInt(2),
        params: {
          recipientAddress: getAddressFromBase32Address(wallets.genesis.summary.address),
          amount: BigInt(112300000),
          data: 'test',
        },
        signatures: expect.any(Array),
        id: expect.any(Object),
      };

      // Act
      await tokensTransferred(data)(dispatch, getState);
      const expectedAction = {
        type: actionTypes.transactionCreatedSuccess,
        data: tx,
      };

      // Assert
      expect(hwManager.signTransactionByHW).not.toHaveBeenCalled();
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });

    it('should dispatch transactionSignError action if there are errors during transaction creation', async () => {
      // Arrange
      const data = {
        fee: NaN,
        moduleCommandID: '2:0',
        sender: {
          publicKey: wallets.genesis.summary.publicKey,
        },
        nonce: '2',
        params: {
          recipient: { address: wallets.genesis.summary.address },
          amount: 112300000,
          data: 'test',
        },
      };
      const transactionError = new Error(
        'The number NaN cannot be converted to a BigInt because it is not an integer'
      );
      loginTypes.passphrase.code = 1;
      jest.spyOn(hwManager, 'signTransactionByHW').mockRejectedValue(transactionError);
      const expectedAction = {
        type: actionTypes.transactionSignError,
        data: transactionError,
      };
      // Act

      await tokensTransferred(data)(dispatch, getStateWithHW);
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });
  });
});
