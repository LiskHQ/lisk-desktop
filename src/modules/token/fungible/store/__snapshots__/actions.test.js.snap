// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions: transactions tokensTransferred should dispatch tokensTransferredSuccess action if there are no errors 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": [TypeError: Cannot read properties of undefined (reading 'split')],
        "type": "TRANSACTION_SIGN_ERROR",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
